projects/lights_control/src/lights_can.c

```diff
@@ -0,0 +1,89 @@
+#include "can.h"
+#include "can_hw.h"
+#include "can_msg_defs.h"
+#include "can_unpack.h"
+#include "event_queue.h"
+#include "status.h"
+
+#include "exported_enums.h"
+#include "lights_can.h"
+#include "lights_events.h"
+
+// RX handler for signal type events.
+static StatusCode prv_rx_signal_handler(const CANMessage *msg, void *context,
+                                        CANAckStatus *ack_reply) {
+  uint8_t signal_type = 0;
+  uint8_t state = 0;
+  status_ok_or_return(CAN_UNPACK_LIGHTS_SIGNAL_TYPES(msg, &signal_type, &state));
+  if (signal_type >= NUM_LIGHTS_SIGNAL_TYPES) {
+    return status_msg(STATUS_CODE_INVALID_ARGS, "Unsupported signals type.");
+  }
+  return event_raise((state) ? LIGHTS_EVENT_SIGNAL_ON : LIGHTS_EVENT_SIGNAL_OFF,
+                     (LightsEventSignalMode)signal_type);
+}
+
+// RX handler for generic lights.
+static StatusCode prv_rx_generic_lights_handler(const CANMessage *msg, void *context,
+                                                CANAckStatus *ack_reply) {
+  LightsCanSettings *settings = (LightsCanSettings *)context;
+  uint8_t light_type = 0;
+  uint8_t state = 0;
+  status_ok_or_return(CAN_UNPACK_LIGHTS_GENERIC_TYPES(msg, &light_type, &state));
+  if (light_type >= NUM_LIGHTS_GENERIC_TYPES) {
+    return status_msg(STATUS_CODE_INVALID_ARGS, "Unsupported generic light type.");
+  }
+  return event_raise((state) ? LIGHTS_EVENT_GPIO_ON : LIGHTS_EVENT_GPIO_OFF,
+                     (LightsEventSignalMode)settings->generic_light_lookup[light_type]);
+}
+
+// RX handler for sync.
+static StatusCode prv_rx_sync_handler(const CANMessage *msg, void *context,
+                                      CANAckStatus *ack_reply) {
+  status_ok_or_return(CAN_UNPACK_LIGHTS_SYNC(msg));
```

Comment:
You realize this does nothing? I'm surprised it compiles.

Reply:
Why not? I added the implementation myself: 
```
#define CAN_UNPACK_LIGHTS_SYNC(msg_ptr) can_unpack_impl_u64((msg_ptr), 0, CAN_UNPACK_IMPL_EMPTY)
```
I saw that `can_unpack_impl_empty()` doesn't do anything, which is why I used `can_unpack_impl_u64` instead. 

P.S. I realize that codegen-tooling will generate the `CAN_UNPACK_LIGHTS_SYNC` in the future, I just manually added it here for this to work.

Reply:
There's no point in unpacking an empty message. Codegen-tooling will use `can_unpack_impl_empty`.